cli/command/service/list.go

```diff
@@ -89,54 +92,97 @@ func runList(dockerCli command.Cli, options listOptions) error {
 
 	servicesCtx := formatter.Context{
 		Output: dockerCli.Out(),
-		Format: NewListFormat(format, options.quiet),
+		Format: NewListFormat(format, opts.quiet),
 	}
-	return ListFormatWrite(servicesCtx, services, info)
+	return ListFormatWrite(servicesCtx, services)
 }
 
-// GetServicesStatus returns a map of mode and replicas
-func GetServicesStatus(services []swarm.Service, nodes []swarm.Node, tasks []swarm.Task) map[string]ListInfo {
-	running := map[string]int{}
-	tasksNoShutdown := map[string]int{}
-
-	activeNodes := make(map[string]struct{})
-	for _, n := range nodes {
-		if n.Status.State != swarm.NodeStateDown {
-			activeNodes[n.ID] = struct{}{}
+// AppendServiceStatus propagates the ServiceStatus field for "services".
+//
+// If API version v1.41 or up is used, this information is already set by the
+// daemon. On older API versions, we need to do some extra requests to get
+// that information. Theoretically, this function can be skipped based on API
+// version, however, some of our unit tests don't set the API version, and
+// there may be other situations where the client uses the "default" version.
+// To take these situations into account, we do a quick check for services
+// that don't have ServiceStatus set, and perform a lookup for those.
+// nolint: gocyclo
+func AppendServiceStatus(ctx context.Context, c client.APIClient, services []swarm.Service) ([]swarm.Service, error) {
+	status := map[string]*swarm.ServiceStatus{}
+	taskFilter := filters.NewArgs()
+	for i, s := range services {
+		switch {
+		case s.ServiceStatus != nil:
+			// Server already returned service-status, so we don't
+			// have to look-up tasks for this service.
+			continue
+		case s.Spec.Mode.Replicated != nil:
+			// For replicated services, set the desired number of tasks;
+			// that way we can present this information in case we're unable
+			// to get a list of tasks from the server.
+			services[i].ServiceStatus = &swarm.ServiceStatus{DesiredTasks: *s.Spec.Mode.Replicated.Replicas}
+			status[s.ID] = &swarm.ServiceStatus{}
+			taskFilter.Add("service", s.ID)
+		case s.Spec.Mode.Global != nil:
+			// No such thing as number of desired tasks for global services
+			services[i].ServiceStatus = &swarm.ServiceStatus{}
+			status[s.ID] = &swarm.ServiceStatus{}
+			taskFilter.Add("service", s.ID)
+		default:
+			// Unknown task type
 		}
 	}
+	if len(status) == 0 {
+		// All services have their ServiceStatus set, so we're done
```

Comment:
Theoretically, we could skip this whole function based on API version, but in situations where the API version is unknown/not set, we would not be propagating `ServiceStatus` (which happens, e.g., in our unit tests). To be on the safe side, the loop above, and this check, make sure that we always check if `ServiceStatus` is present, and otherwise fall back to the old behaviour of fetching tasks and nodes to get this information.